### PR TITLE
Gmsh doc

### PR DESCRIPTION
**Pull request by jeannejoachim**
# Description of the solution

Add GMSH help page on documentation

### CODE REVIEW

File: doc/source/tools/gmsh/gmsh.rst [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+Basic:
+
+1. (optional) ``Tools > Options > Mesh`` and ``General`` panel, check ``Recombine all triangular meshes``: generate a quad mesh.
+2. (optional) In the same panel, change ``Min/Max element size`` to have smaller/bigger elements, therefor a finer/coarser mesh.
+3. ``Left pannel: Modules > Mesh > 2D`` or ``3D``: create the mesh
+4. ``Tools > Statistics``: check that the mesh is generated appropriately (by default, triangles for 2D and hexahedra for 3D)
+5. (optional) ``Left pannel: Modules > Mesh > Refine by splitting``: refine the mesh (beware, it takes more and more time for each refinement)
+6. ``Left pannel: Modules > Mesh > Save``: save the mesh in a ``.msh`` file, to be used in Lethe (see :doc:`../../parameters/cfd/mesh`)
+
+By following all the previous steps, the mesh generated looks like bellow.
```
Comment (tonielgeitani) [inline doc/source/tools/gmsh/gmsh.rst:245]:
```suggestion
By following all the previous steps, the mesh generated looks like below.
```

**Review by tonielgeitani — APPROVED** [review]
Very nice explanation. I suggest adding a sentence with a link to a lethe-util folder where we can have all our gosh .geo files. For example, I can put my spouted bed gosh file and so on.

File: doc/source/tools/gmsh/gmsh.rst [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+
+""""""""""""""""""""""""""
+Built-in kernel
+""""""""""""""""""""""""""
+It is quite easy to create a ``.geo`` file directly by coding line by line the geometry that you want to produce.
+
+.. tip::
+	Using a Built-in kernel gives you full control over the geometry creation, which is necessary to create a :ref:`structured mesh` (transfinite) mesh. However, it usually takes more time than using functions available in :ref:`opencascade kernel`.
+
+1. Set the points that delimits the shape being traced. The attribut ``Point`` is essential to set a point of the shape and to specify its coordinates.
```
Comment (CaroleAnneDaunais) [inline doc/source/tools/gmsh/gmsh.rst:57]:
```suggestion
1. Set the points that delimits the shape being traced. The attribute ``Point`` is essential to set a point of the shape and to specify its coordinates. 
```

File: doc/source/tools/gmsh/gmsh.rst [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+.. code-block::
+
+	// Curve Loop(<id>) = {<id of line>, ...}
+	Curve Loop(1) = {1, 2, 3, 4};
+	Curve Loop(2) = {5, 6};
+	// Plane Surface(<id>) = {<id of curve loop>, ...}
+	Plane Surface(1) = {1, 2};
+	
+.. tip::
+	All the lines of code can be directly made with the GUI of gmsh with some clicks and keyboards shortcuts.
```
Comment (CaroleAnneDaunais) [inline doc/source/tools/gmsh/gmsh.rst:112]:
```suggestion
	All the lines of code can be directly made with the GUI of gmsh with some clicks and keyboard shortcuts.
```

File: doc/source/tools/gmsh/gmsh.rst [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+
+.. _opencascade kernel:
+
+""""""""""""""""""""""""""
+OpenCASCADE kernel
+""""""""""""""""""""""""""
+In the GMSH geometry section of the GMSH GUI (see ``Left pannel: Modules > Geometry > Elementary entities > Add``), you can add directly multiple 2D or 3D common geometries with a simple click thanks to OpenCASCADE kernel. GMSH will automatically open a window where you can easily set the characteristic lenghts of the geometry, and update the ``.geo`` file.
+
+.. warning::
+	Always save your ``.geo`` file in your text editor before modifying it through the GMSH GUI. If you modified the ``.geo`` file without saving it, GMSH will not update it.
```
Comment (CaroleAnneDaunais) [inline doc/source/tools/gmsh/gmsh.rst:122]:
```suggestion
	Always save your ``.geo`` file in your text editor before modifying it through the GMSH GUI. If you modify the ``.geo`` file without saving it, GMSH will not update it. 
```

File: doc/source/tools/gmsh/gmsh.rst [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+1. Select the ``Disk`` geometry available with OpenCASCADE, set the radius to 1 (for X and Y) and center it at :math:`(x,y)=(0,0)`. 
+2. Select the ``Rectangle`` geometry, set the length (``DX``) to :math:`20`, the width (``DY``) to :math:`10` and the left bottom corner to :math:`(x,y)=(-5,-5)`.
+    
+.. note::
+	If you click on ``Edit script``, you will see that the OpenCASCADE kernel has been added to the code as ``SetFactory("OpenCASCADE");``. 
+
+	The rectangle is set with ``Rectangle(1) = {-5, -5, 0, 20, 10, 0};`` and the circle with ``Disk(2) = {0, 0, 0, 1, 1};``.
+	
+.. tip::
+	The ``Disk`` and ``Rectangle`` are already considered as surfaces in gmsh, so no need to pass from points, to curves and then surfaces.
```
Comment (CaroleAnneDaunais) [inline doc/source/tools/gmsh/gmsh.rst:135]:
```suggestion
	The ``Disk`` and ``Rectangle`` are already considered as surfaces in gmsh, there is therefore no need to pass from points, to curves and then surfaces. 
```

File: doc/source/tools/gmsh/gmsh.rst [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+Import CAD file
+""""""""""""""""""""""""""
+Importing CAD files (``.step`` or ``.stp`` format) can be particularly convenient for more complex fluid geometries (like pipes) or immersed solids (like an impeller):
+
+.. hint::
+  In the case of immersed solids, use a simplified CAD file of the outer shell of the solid, e.g. without any screws or bolts or threads.
+
+1. ``File > New...``: create a new .geo file (can use OpenCASCADE or Built-in kernel)
+2. ``Files > Merge...``: merge the CAD file (``.step`` or ``.stp`` format) with GMSH
+3. ``Tools > Statistics``: check that the geometry is loaded (point, curves, surfaces, and if 3D volumes)
```
Comment (CaroleAnneDaunais) [inline doc/source/tools/gmsh/gmsh.rst:154]: What does "and if 3D volumes" mean?

File: doc/source/tools/gmsh/gmsh.rst [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+
+.. seealso::
+  You can find a step-by-step video `here <https://www.youtube.com/watch?v=e7zA3joOWX8>`_, with very useful tools as how to inspect your mesh.
+
+--------------------------
+Physical group
+--------------------------
+
+.. warning::
+	This step is essential. Physical group are used to identify the boundary conditions.
```
Comment (CaroleAnneDaunais) [inline doc/source/tools/gmsh/gmsh.rst:164]:
```suggestion
	This step is essential. Physical groups are used to identify the boundary conditions.
```

File: doc/source/tools/gmsh/gmsh.rst [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+6. ``Left pannel: Modules > Mesh > Save``: save the mesh in a ``.msh`` file, to be used in Lethe (see :doc:`../../parameters/cfd/mesh`)
+
+By following all the previous steps, the mesh generated looks like bellow.
+
+.. image:: images/unstructured.png
+    :alt: 2D mesh with quads
+    :align: center
+    :name: unstructured mesh
+    
+Attractors can also be used to refine the mesh towards specific edges or surfaces. In this example, attractors could be interesting if the mesh needs to be finer around the sphere. Attractors can only be added by code with the ``Field`` attribut.
```
Comment (CaroleAnneDaunais) [inline doc/source/tools/gmsh/gmsh.rst:252]: Should there be a subsection title called Attractors?

File: doc/source/tools/gmsh/gmsh.rst [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+
+.. _structured mesh:
+
+""""""""""""""""""""""""""
+Structured - Quad mesh
+""""""""""""""""""""""""""
+.. warning::
+	The ``.geo`` file must be built with the :ref:`built-in kernel`.
+
+Creating a structured quad-mesh usually take a lot more time than an unstructured quad-mesh, but provides a full control on the mesh generation. To do so:
```
Comment (CaroleAnneDaunais) [inline doc/source/tools/gmsh/gmsh.rst:308]:
```suggestion
Creating a structured quad-mesh usually takes a lot more time than an unstructured quad-mesh, but provides a full control on the mesh generation. To do so:
```

File: doc/source/tools/gmsh/gmsh.rst [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+
+1. Create the geometry accordingly and add construction elements where needed
+
+.. tip::
+	Converting an unstructured mesh to a structured mesh usually required rewritting a good part of the geometry. Begin by drawing by-hand and check that each of your surfaces have only four points.
```
Comment (CaroleAnneDaunais) [inline doc/source/tools/gmsh/gmsh.rst:313]:
```suggestion
	Converting an unstructured mesh to a structured mesh usually requires rewritting a good part of the geometry. Begin by drawing by-hand and check that each of your surfaces have only four points.
```

Comment (CaroleAnneDaunais) [inline doc/source/tools/gmsh/gmsh.rst:313]:
```suggestion
	Converting an unstructured mesh to a structured mesh usually required rewritting a good part of the geometry. Begin by drawing by-hand and check that all of your surfaces have only four points each.
```

File: doc/source/tools/gmsh/gmsh.rst [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+.. code-block::
+
+	// Recombine Surface {<id>}
+
+5. Generate the mesh
+
+.. seealso::
+  You can find a step-by-step `video <https://youtu.be/1A-b84kloFs?t=316>`_, with a similar geometry (cylinder in flow). 
+
+A mesh can also be partially structured, to better account for a boundary layer for instance: see the ``.geo`` file provided with the :doc:`../../examples/incompressible-flow/2d-transient-around-ahmed-body/2d-transient-around-ahmed-body` example.
```
Comment (CaroleAnneDaunais) [inline doc/source/tools/gmsh/gmsh.rst:334]:
```suggestion
A mesh can also be partially structured, to better encounter for a boundary layer for instance: see the ``.geo`` file provided with the :doc:`../../examples/incompressible-flow/2d-transient-around-ahmed-body/2d-transient-around-ahmed-body` example. 
```

File: doc/source/tools/gmsh/gmsh.rst [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+
+		//Parameters
+		L = 5; 			//length
+		C45 = Cos(45*Pi/180); 	//cosine of 45 degrees
+		esf = 1.2; 		//element size factor
+
+		//Points
+		Point(1) = {L, L*C45, 0, esf};
+
+- Use the ``Visibility`` options to get the ID of an element or a physical group easily on the GUI:
```
Comment (CaroleAnneDaunais) [inline doc/source/tools/gmsh/gmsh.rst:354]:
```suggestion
- Use the ``Visibility`` option to get the ID of an element or a physical group easily on the GUI: 
```

File: doc/source/tools/gmsh/gmsh.rst [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+	* ``Tools > Options > Mesh > Tab: Visibility``
+	* Check the adequate boxes (for example ``1D element labels`` for points, etc.) 
+	* Choose the label type in the drop-down menu ``Label type`` (for example ``Elementary entity tag``).
+
+- You can define a range of elements to group, and if an index does not exist it is simply not considered. Use it to define ``Physical Surface`` (2D) or ``Physical Volume`` (3D) for the whole geometry more easily.
+	* For example: ``Physical Surface(1) = {1:200};``
+
+- Click on the grey bar at the bottom of the software interface to see all the logs, errors and warnings.
+
+- Verify your mesh in: ``Tools > Statistics > Mesh``. In particular pay attention that you only have one type of elements. If not, changing the mesh refinement can help removing unwanted triangles in a quad mesh.
```
Comment (CaroleAnneDaunais) [inline doc/source/tools/gmsh/gmsh.rst:364]:
```suggestion
- Verify your mesh in: ``Tools > Statistics > Mesh``. In particular, make sure that you only have one type of elements. If not, changing the mesh refinement can help removing unwanted triangles in a quad mesh.
```

**Review by CaroleAnneDaunais — COMMENTED** [review]
Excellent work :)

File: doc/source/tools/gmsh/gmsh.rst (hunk 8 above)
Comment (blaisb) [reply doc/source/tools/gmsh/gmsh.rst:252]: Agreed. Or maybe local mesh refinement? Because this is what Attractors achieve.

**Review by blaisb — CHANGES_REQUESTED** [review]
Only a single comment on my side so far. This is very, very cool work. Good job.

File: doc/source/tools/gmsh/gmsh.rst (hunk 6 above)
Comment (bibeauv) [reply doc/source/tools/gmsh/gmsh.rst:154]: Volumes if the geometry is in 3D :)

File: doc/source/tools/gmsh/gmsh.rst (hunk 8 above)
Comment (bibeauv) [reply doc/source/tools/gmsh/gmsh.rst:252]: Good idea!

File: doc/source/tools/gmsh/gmsh.rst (hunk 12 above)
Comment (bibeauv) [reply doc/source/tools/gmsh/gmsh.rst:354]: It is a tool actually, so I think I am gonna go with tool instead of option.

Comment (bibeauv) [reply doc/source/tools/gmsh/gmsh.rst]: Never mind, it is an option.

**Review by CaroleAnneDaunais — APPROVED** [review]

**Review by blaisb — APPROVED** [review]

**Review by bibeauv — APPROVED** [review]
Ready to merge :)

**Pull request merged**